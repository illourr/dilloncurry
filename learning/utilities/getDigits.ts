function getDigits(num: number): number[] {
  if (!Number.isSafeInteger(num)) {
    throw new RangeError("Value is not a safe integer!");
  }

  if (num === 0) {
    return [0];
  }

  const digits = [];
  let absNum = Math.abs(num);
  while (absNum > 0) {
    // get the digits
    const digit = absNum % 10;

    // save the digit
    digits.unshift(Number(digit));

    // remove the digit
    absNum = Math.floor(absNum / 10);
  }
  return digits;
}

function getDigitsBigInt(num: bigint): number[] {
  if (num === 0n) {
    return [0];
  }

  const digits = [];
  let tempNum = num;
  while (tempNum > 0n) {
    // get the digits
    const digit = tempNum % 10n;

    // save the digit
    digits.unshift(Number(digit));

    // remove the digit
    tempNum = tempNum / 10n;
  }
  return digits;
}

export { getDigits, getDigitsBigInt };
