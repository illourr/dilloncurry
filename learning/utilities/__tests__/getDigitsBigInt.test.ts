import { test, expect } from "bun:test";
import { getDigitsBigInt } from "../getDigits";

test("big ole bigints work", () => {
  expect(
    getDigitsBigInt(
      30392483204832904382049320498320483248230483209480924809328409328409832409832490329482348n,
    ),
  ).toEqual([
    3, 0, 3, 9, 2, 4, 8, 3, 2, 0, 4, 8, 3, 2, 9, 0, 4, 3, 8, 2, 0, 4, 9, 3, 2,
    0, 4, 9, 8, 3, 2, 0, 4, 8, 3, 2, 4, 8, 2, 3, 0, 4, 8, 3, 2, 0, 9, 4, 8, 0,
    9, 2, 4, 8, 0, 9, 3, 2, 8, 4, 0, 9, 3, 2, 8, 4, 0, 9, 8, 3, 2, 4, 0, 9, 8,
    3, 2, 4, 9, 0, 3, 2, 9, 4, 8, 2, 3, 4, 8,
  ]);
});
