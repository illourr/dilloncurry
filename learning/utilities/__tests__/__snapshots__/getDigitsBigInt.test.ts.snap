// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`big ole bigints work 1`] = `
[
  3,
  0,
  3,
  9,
  2,
  4,
  8,
  3,
  2,
  0,
  4,
  8,
  3,
  2,
  9,
  0,
  4,
  3,
  8,
  2,
  0,
  4,
  9,
  3,
  2,
  0,
  4,
  9,
  8,
  3,
  2,
  0,
  4,
  8,
  3,
  2,
  4,
  8,
  2,
  3,
  0,
  4,
  8,
  3,
  2,
  0,
  9,
  4,
  8,
  0,
  9,
  2,
  4,
  8,
  0,
  9,
  3,
  2,
  8,
  4,
  0,
  9,
  3,
  2,
  8,
  4,
  0,
  9,
  8,
  3,
  2,
  4,
  0,
  9,
  8,
  3,
  2,
  4,
  9,
  0,
  3,
  2,
  9,
  4,
  8,
  2,
  3,
  4,
  8,
]
`;
